### PR TITLE
Implementing OPE Estimators for Multiple Loggers

### PR DESCRIPTION
**Pull request by usaito**
## Overview
- Implement OPE estimators for the case with multiple logging policies, including Naive IPW, Balanced IPW, Weighted IPW, Naive DR, Weighted DR, and Balanced DR. These estimators are proposed in the reference papers.
- Add corresponding tests

## Minor points
- rename `obp.dataset.SyntheticBanditDatasetWithMultiLoggers` to `obp.dataset.SyntheticMultiLoggersBanditDataset` 
- some refactoring

## References
- Aman Agarwal, Soumya Basu, Tobias Schnabel, and Thorsten Joachims.
 "Effective Evaluation using Logged Bandit Feedback from Multiple Loggers.", 2018.
https://arxiv.org/abs/1703.06180

- Nathan Kallus, Yuta Saito, and Masatoshi Uehara.
"Optimal Off-Policy Evaluation from Multiple Logging Policies.", 2021.
https://arxiv.org/abs/2010.11002

### CODE REVIEW

File: obp/ope/estimators_multi.py [hunk 1]
```diff
@@ -0,0 +1,1806 @@
...
+        )
+
+
+@dataclass
+class MultiLoggersWeightedDoublyRobust(MultiLoggersNaiveDoublyRobust):
+    """Multi-Loggers Naive Doubly Robust (Multi-Weighted-DR) Estimator.
+
+    Note
+    -------
+    This estimator is called Prevision Weighted DR in Kallus et al.(2021).
```
Comment (fullflu) [inline obp/ope/estimators_multi.py:1494]:
```suggestion
    This estimator is called Precision Weighted DR in Kallus et al.(2021).
```

File: obp/ope/estimators_multi.py [hunk 2]
```diff
@@ -0,0 +1,1806 @@
...
+        # weight clipping
+        if isinstance(iw, np.ndarray):
+            iw = np.minimum(iw, self.lambda_)
+
+        unique_stratum_idx, n_data_strata = np.unique(stratum_idx, return_counts=True)
+        var_k = np.zeros(unique_stratum_idx.shape[0])
+        for k in unique_stratum_idx:
+            idx_ = stratum_idx == k
+            var_k[k] = np.var(reward[idx_] * iw[idx_])
+        weight_k = n / (var_k * np.sum(n_data_strata / var_k))
```
Comment (fullflu) [inline obp/ope/estimators_multi.py:691]: [ask]
Is this a typo?

```suggestion
        weight_k = n_data_strata / (var_k * np.sum(n_data_strata / var_k))
```

File: obp/ope/estimators_multi.py [hunk 3]
```diff
@@ -0,0 +1,1806 @@
...
+
+        unique_stratum_idx, n_data_strata = np.unique(stratum_idx, return_counts=True)
+        var_k = np.zeros(unique_stratum_idx.shape[0])
+        for k in unique_stratum_idx:
+            idx_ = stratum_idx == k
+            var_k[k] = np.var(
+                estimated_rewards[idx_]
+                + iw[idx_] * (reward[idx_] - q_hat_factual[idx_])
+            )
+        weight_k = n / (var_k * np.sum(n_data_strata / var_k))
```
Comment (fullflu) [inline obp/ope/estimators_multi.py:1611]: [ask]
Is this a typo?

```suggestion
        weight_k = n_data_strata / (var_k * np.sum(n_data_strata / var_k))
```

**Review by fullflu — COMMENTED** [review]
1st Review of `estimators_multi.py`

File: obp/ope/estimators_multi.py (hunk 2 above)
Comment (usaito) [reply obp/ope/estimators_multi.py:691]: @fullflu Thanks! Actually, I think my implementation is correct. In the following definition, there is no (1/n), but in our implementation, we always take the average of the estimated rewards in `estimate_policy_value`. So, we have to multiply the estimated_rewards by `n` in advance for this estimator.

スクリーンショット 2022-02-09 2 29 51

Comment (fullflu) [reply obp/ope/estimators_multi.py:691]: OK, I agree with you!!

File: obp/dataset/synthetic_multi.py [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
 
 @dataclass
-class SyntheticBanditDatasetWithMultiLoggers(SyntheticBanditDataset):
+class SyntheticMultiLoggersBanditDataset(SyntheticBanditDataset):
```
Comment (fullflu) [inline obp/dataset/synthetic_multi.py:23]: [nits]

`rhos: List[float]` should be `rhos: List[Union[int, float]]` (comment)

**Review by fullflu — COMMENTED** [review]
I finished the entire review!

I also checked the whole lines of `synthetic_multi.py` (LGTM).

**Pull request merged**